crates/auth/src/types.rs

```diff
@@ -0,0 +1,142 @@
+pub struct User {
+    pub auth_type: AuthType,
+    pub token: AccessToken,
+}
+
+impl User {
+    /// Decrypt an untrusted user token to guarantee it was not modified by the user.
+    pub fn from_encrypted_str(string: &str, key: &RsaPrivateKey) -> Result<Self, Error> {
+        let bytes = base64::decode(string).map_err(Error::TokenDecode)?;
+        let bytes = key
+            .decrypt(PaddingScheme::new_pkcs1v15_encrypt(), &bytes)
+            .map_err(Error::TokenDecrypt)?;
+        bincode::deserialize(&bytes).map_err(Error::TokenDeserialize)
+    }
+
+    /// Encrypt the session so that it can be securely stored by the user.
+    pub fn encrypt(&self, key: &RsaPrivateKey) -> Result<String, Error> {
+        let bytes = bincode::serialize(self).map_err(Error::TokenSerialize)?;
+        key.encrypt(
+            &mut OsRng,
+            PaddingScheme::new_pkcs1v15_encrypt(),
+            &bytes[..],
+        )
+        .map_err(Error::TokenEncrypt)
+        .map(base64::encode)
+    }
+
+    /// Validate if a session should be considered active.
+    pub async fn validate(&self) -> Result<String, Error> {
+        match self.auth_type {
+            AuthType::GitHub => {
+                let client = reqwest::Client::new();
+
+                #[derive(Deserialize)]
+                struct GitHubUser {
+                    login: String,
+                }
+
+                #[derive(Deserialize)]
+                struct GitHubError {
+                    message: String,
+                }
+
+                let body = client
+                    .get("https://api.github.com/user")
+                    .header(USER_AGENT, "drawbridge")
+                    .bearer_auth(self.token.secret())
+                    .send()
+                    .await
+                    .map_err(Error::Request)?
+                    .text()
+                    .await
+                    .map_err(Error::Request)?;
+
+                let user = serde_json::from_str::<GitHubUser>(&body).map_err(|_| {
+                    match serde_json::from_str::<GitHubError>(&body) {
+                        Err(e) => Error::Serde(e.to_string()),
+                        Ok(error) => Error::OAuth(error.message),
+                    }
+                })?;
+
+                Ok(user.login)
+            }
+        }
+    }
+}
+
+#[async_trait]
+impl<B> FromRequest<B> for User
+where
+    B: Send,
+{
+    type Rejection = AuthRedirect;
+
+    async fn from_request(req: &mut RequestParts<B>) -> Result<Self, Self::Rejection> {
+        let key = Extension::<RsaPrivateKey>::from_request(req).await.unwrap();
+        let cookies = TypedHeader::<headers::Cookie>::from_request(req)
+            .await
+            .map_err(|e| match *e.name() {
+                header::COOKIE => match e.reason() {
+                    TypedHeaderRejectionReason::Missing => AuthRedirect,
+                    _ => panic!("unexpected error getting Cookie header(s): {}", e),
```

Comment:
We should log the error and return a 500 to the user, but we can't panic

Reply:
Definitely. I have a fix for this locally and some other fixes I will push sometime today.